graph-builder/src/registry.rs

```diff
@@ -42,55 +42,148 @@ impl Into<cincinnati::Release> for Release {
+                insecure: Registry::insecure_scheme(&scheme)?,
+                scheme,
+                host: capture["host"].to_string(),
+                ..Default::default()
+            });
+        };
+
+        if let Some(capture) = regex::Regex::new(
+            // match h.o.s.t:port
+            r"^(?P<host>([0-9a-zA-Z\-]+\.)+([0-9a-zA-Z]+)):(?P<port>[0-9]+)$",
+        )
+        .context("could not compile regex pattern")?
+        .captures(src)
+        {
+            return Ok(Registry {
+                host: capture["host"].to_string(),
+                port: Some(
+                    capture["port"]
+                        .parse()
+                        .expect("could not parse port as a number"),
+                ),
+                ..Default::default()
+            });
+        };
+
+        if let Some(capture) = regex::Regex::new(
+            // match h.o.s.t
+            r"^(?P<host>([0-9a-zA-Z\-]+\.)*([0-9a-zA-Z]+))$",
+        )
+        .context("could not compile regex pattern")?
+        .captures(src)
+        {
+            return Ok(Registry {
+                host: capture["host"].to_string(),
+                ..Default::default()
+            });
+        };
+
+        bail!("unsupported registry format {}", src)
+    }
+
+    pub fn try_new(scheme: String, host: String, port: Option<u16>) -> Fallible<Self> {
+        Ok(Registry {
+            host,
+            port,
+            insecure: Self::insecure_scheme(&scheme)?,
+            scheme,
+        })
+    }
+
+    pub fn host_port_string(&self) -> String {
+        format!(
+            "{}{}",
+            self.host,
+            if let Some(port) = self.port {
+                format!(":{}", port)
+            } else {
+                "".to_string()
+            }
+        )
+    }
+
+    fn insecure_scheme(scheme: &str) -> Fallible<bool> {
+        match scheme {
+            "https" | "docker" => Ok(false),
+            "http" => Ok(true),
+            scheme => bail!("unsupported url scheme specified '{}'", scheme),
+        }
+    }
 }
 
 pub fn read_credentials(
     credentials_path: Option<&PathBuf>,
-    registry: &str,
+    registry_host: &str,
 ) -> Result<(Option<String>, Option<String>), Error> {
-    credentials_path.clone().map_or(Ok((None, None)), |path| {
-        Ok(
-            dkregistry::get_credentials(File::open(&path)?, trim_protocol(registry))
-                .map_err(|e| format_err!("{}", e))?,
-        )
+    credentials_path.map_or(Ok((None, None)), |path| {
+        let file = File::open(&path).context(format!("could not open '{:?}'", path))?;
+
+        Ok(dkregistry::get_credentials(file, &registry_host).map_err(|e| format_err!("{}", e))?)
     })
 }
 
 pub fn authenticate_client(
-    client: &mut dkregistry::v2::Client,
+    dclient: &mut dkregistry::v2::Client,
     login_scope: String,
 ) -> impl Future<Item = &dkregistry::v2::Client, Error = dkregistry::errors::Error> {
-    client
-        .is_v2_supported()
```

Comment:
This is missing in the new revision, just checking if it got dropped on purpose.

Reply:
I originally dropped this on purpose as I think that the the v2 client should throw an error if it can't talk to the registry as expected. I do understand that this check might be in scope of the registry module here so I'll bring it back in.